### ISSUE TITLE
Extremely poor performance on ER-X

### ISSUE THREAD

**Issue by smartypantsuk** [WireGuard/wireguard-vyatta-ubnt#151]
Hi,

I've just upgraded my phone to Android 12 and found out that my old VPN will no longer work, so I've installed Wireguard which installed fine and I can connect very quickly. 
Setup is simple:
ER-X with Fibre modem on port eth0 (pppoe)
Access point on eth2
The usual home wifi devices and one media server on local network.
3 vlans.

My internet is pretty poor, aprox 35Mb/8Mb, due to be upgraded soon luckily.
HW Offloading is enabled though I've tried disabling with no change.

Not done any iperf testing, but the internet speed test while directly connected via via to my network is around 35Mb, but through the vpn connection it is only around 6Mb.... I usually have an always on vpn connection to home when I leave the house so that I can access all my security cams and smart home stuff. 6Mb is useless!

I have no idea where to start looking at this, so my guess is either a firewall or possible mtu issue?
What is the process to diagnose this please?

Thanks

**Comment by dc361** [comment]
Just a guess but I'm thinking MTU.  The MTU on my IOS client and my ER4 is 1420 and I get over 80% of my ISP bandwidth over the Wireguard connection.

**Comment by smartypantsuk** [comment]
> Just a guess but I'm thinking MTU. The MTU on my IOS client and my ER4 is 1420 and I get over 80% of my ISP bandwidth over the Wireguard connection.

While I've not forced any mtu setting, I'm assuming the defaults on both the client and server are 1420, unless there is a bug.
I could try forcing mtu I guess.

**Comment by dc361** [comment]
I agree but it is one of the easiest settings to try and I know when I originally tried wireguard on edgerouter the default MTU was 1500.

The other thing to check is that you have enabled hwnat (which is what the ER-X uses) and not IPV4/IPV6 forwarding which is what the ERL, ER4, etc use.

**Comment by smartypantsuk** [comment]
> I agree but it is one of the easiest settings to try and I know when I originally tried wireguard on edgerouter the default MTU was 1500.
> 
> The other thing to check is that you have enabled hwnat (which is what the ER-X uses) and not IPV4/IPV6 forwarding which is what the ERL, ER4, etc use.

Okay so I've forced 1420 on both edgerouter and client, I've also tried other variations around the 1300s but I see no improvements whatsoever. 

Re: hwnat. 
My understanding is that on the ER-X, you can only enable hw offload and ipsec offload as options. If you try to enable them individually it errors, but I thought when you enabled hw offload it enables hwnat too?

**Comment by smartypantsuk** [comment]
Further to me previous post.

I tethered my laptop to my phone. My phone was connected to 4G connection working around 45Mbps.
I installed the Wireguard VPN on my laptop and connected to my network through the vpn via tethered 4G phone.
I then ran a speedtest and got 15.5Mbps.

While not up to the direct network connected 35Mbps, it's nearly three times faster than when I use the Android Wireguard VPN app through the same 4G connection.

Not sure whether this points to a problem with the Wireguard Android app or MTU

**Comment by Salohcins** [comment]
I think I might be missing something here...  Won't your maximum available bandwidth be capped/determined by your upload speed to your remote device?

For comparison sake, what speed do you get on your laptop when you're on the 4G connection but _not_ connected to the VPN?

To me it sounds like your phone is doing the correct thing (6Mb ~= 8Mb), and the laptop isn't.  Really love to know how you go, as I'm currently considering using my router for Wireguard also.

**Comment by smartypantsuk** [comment]
> I think I might be missing something here... Won't your maximum available bandwidth be capped/determined by your upload speed to your remote device?
> 
> For comparison sake, what speed do you get on your laptop when you're on the 4G connection but _not_ connected to the VPN?
> 
> To me it sounds like your phone is doing the correct thing (6Mb ~= 8Mb), and the laptop isn't. Really love to know how you go, as I'm currently considering using my router for Wireguard also.

Yes that's the problem. 
Sorry I forgot to update the post. 

I completely forgot that upload speed at the destination end is relevant when the client is behind another internet connection.  It's great to know I wasted everyone's time 😆 .

So for the record,  nothing wrong.  I need to wait for my internet provider to upgrade us all to faster  fibre. 

Thanks

**Comment by smartypantsuk** [comment]
Yes that was the problem. 
I was having a very dumb moment. 
If the client is behind a different Internet connection,  the upload speed of the destination network is the limiting factor. 

Sorry to waste your time,  everything seems fine. 
Rubbish for me but fine 😄
